### PR TITLE
Introduce `rb_memsearch_with_char_size` function

### PR DESCRIPTION
**Pull request by S-H-GAMELINKS**
`rb_memsearch_wchar` and `rb_memsearch_qchar` functions has almost same code.

I thought more better that introduce `rb_memsearch_with_char_size` function and merge these functions to `rb_memsearch_with_char_size`.

### CODE REVIEW

File: re.c [hunk 1]
```diff
@@ -220,11 +220,16 @@ rb_memsearch_qs_utf8(const unsigned char *xs, long m, const unsigned char *ys, l
...
 
+enum char_size
+{
+    WCHAR_SIZE = 2,
+    QCHAR_SIZE = 4
+};
+
 static inline long
-rb_memsearch_wchar(const unsigned char *xs, long m, const unsigned char *ys, long n)
+rb_memsearch_with_char_size(const unsigned char *xs, long m, const unsigned char *ys, long n, enum char_size char_size)
```
Comment (nobu) [inline re.c:230]: As the `char_size` was just intended a local constant, you don't have to keep them.
Just literals would be enough.
```suggestion
rb_memsearch_with_char_size(const unsigned char *xs, long m, const unsigned char *ys, long n, int char_size)
```

File: re.c [hunk 2]
```diff
@@ -220,11 +220,16 @@ rb_memsearch_qs_utf8(const unsigned char *xs, long m, const unsigned char *ys, l
     return -1;
 }
 
+enum char_size
+{
+    WCHAR_SIZE = 2,
+    QCHAR_SIZE = 4
+};
+
```
Comment (nobu) [inline re.c:228]:
```suggestion
```

File: re.c [hunk 3]
```diff
@@ -234,16 +239,15 @@ rb_memsearch_wchar(const unsigned char *xs, long m, const unsigned char *ys, lon
 }
 
 static inline long
-rb_memsearch_qchar(const unsigned char *xs, long m, const unsigned char *ys, long n)
+rb_memsearch_wchar(const unsigned char *xs, long m, const unsigned char *ys, long n)
 {
-    const unsigned char *x = xs, x0 = *xs, *y = ys;
-    enum {char_size = 4};
+    return rb_memsearch_with_char_size(xs, m, ys, n, WCHAR_SIZE);
```
Comment (nobu) [inline re.c:244]:
```suggestion
    return rb_memsearch_with_char_size(xs, m, ys, n, 2);
```

File: re.c [hunk 4]
```diff
@@ -234,16 +239,15 @@ rb_memsearch_wchar(const unsigned char *xs, long m, const unsigned char *ys, lon
...
 
-    for (n -= m; n >= 0; n -= char_size, y += char_size) {
-        if (x0 == *y && memcmp(x+1, y+1, m-1) == 0)
-            return y - ys;
-    }
-    return -1;
+static inline long
+rb_memsearch_qchar(const unsigned char *xs, long m, const unsigned char *ys, long n)
+{
+    return rb_memsearch_with_char_size(xs, m, ys, n, QCHAR_SIZE);
```
Comment (nobu) [inline re.c:250]:
```suggestion
    return rb_memsearch_with_char_size(xs, m, ys, n, 4);
```

**Review by nobu — CHANGES_REQUESTED** [review]

**Comment by S-H-GAMELINKS** [comment]
@nobu 
Thanks for your review!
I'd fixed.

**Review by nobu — APPROVED** [review]

**Pull request merged**